src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.HandleObject.cs

```diff
@@ -90,10 +116,49 @@ private static void HandleEndObject(ref ReadStack state)
             }
             else
             {
+                // Set directly when handling preserved array
+                bool setPropertyDirectly = state.Current.IsPreservedArray;
                 state.Pop();
 
-                ApplyObjectToEnumerable(value, ref state);
+                ApplyObjectToEnumerable(value, ref state, setPropertyDirectly);
             }
         }
+
+        private static object GetPreservedArrayValue(ref ReadStack state)
+        {
+            // Preserved JSON arrays are wrapped into JsonPreservedReference<T> where T is the original type of the enumerable
+            // and Values is the actual enumerable instance being preserved.
+            JsonPropertyInfo info = state.Current.JsonClassInfo.PropertyCache["Values"];
```

Comment:
what `enumerable` do you refer to? I didn't quite understand this comment.

Reply:
I'm referring to the preserved array. As we discussed offline, it is possible to store and get the preserved array using `.GetValueAsObject` and `.SetValueAsObject`, without the need for the `JsonPreservedReference<T>`. Furthermore, nested preserved references can be looked up in the reference dictionary.

Since we are keeping `JsonPreservedReference<T>` (to support JsonPath for exceptions I believe), this comment is resolved.

Reply:
This suggestion will mostly be considered for https://github.com/dotnet/runtime/issues/1783 which is about refactoring internal implementation.